numpy/core/src/umath/ufunc_object.c

```diff
@@ -6013,9 +6045,9 @@ NPY_NO_EXPORT PyTypeObject PyUFunc_Type = {
     0,                                          /* tp_getattro */
     0,                                          /* tp_setattro */
     0,                                          /* tp_as_buffer */
-    Py_TPFLAGS_DEFAULT,                         /* tp_flags */
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,    /* tp_flags */
     0,                                          /* tp_doc */
-    0,                                          /* tp_traverse */
+    (traverseproc)ufunc_traverse,               /* tp_traverse */
```

Comment:
I think you also need to set `tp_del` to `PyObject_GC_Del`, but I'm not certain.

Reply:
`tp_free` is set to `PyObject_GC_Del`. `tp_del` [seems to be for internal use only and obsolete](https://bugs.python.org/issue4934).